WcaOnRails/app/jobs/clear_connected_payment_integrations.rb

```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class ClearConnectedPaymentIntegrations < WcaCronjob
+  DELAY_IN_DAYS = 21
+
+  def perform
```

Comment:
I feel like there are more efficient solutions.
1. It is unnecessary to iterate over ALL competitions that ended more than 21 days ago. You will simply blow up the list of competitions more and more over time. Instead, you could just iterate over competitions that are more than X days old _and_ have connected accounts in the first place.
2. Can you have a disconnect that "fails silently" i.e. does nothing if the to-be-disconnected account type was never connected in the first place? Then you can simply iterate over all supported connections and "try to disconnect" them without having to run queries checking whether they're connected.
3. If you really must iterate over lots of competitions (see 1), then use `find_each` instead of `each`. This only works on Rails models and not in general Ruby, but it batches the underlying SQL query _much_ more efficiently.

Bonus idea if you want to feel super pro: What does the `disconnect` method actually do? If it just sets a flag on one table row or just issues a delete, you can re-implement all of this with one simple statement:

```ruby
CompetitionPaymentIntegration.includes(:competition).where(competition: { end_date: ...(DELAY_IN_DAYS.days.ago) }).update_all(is_active: false)
```

Reply:
Great input, thanks! I've updated the code to (I think) only look at comps older than 21 days with a competition payment integration - I think it's happening in 2 calls though, there might be a way to do it in 1. 

Then I'm still iterating through all results and applying `disconnect_all` - I think your super pro solution is more efficient, but hopefully we're dealing with few enough records by that point that efficiency isn't _that_ bad? I'm opting to stick with `disconnect_all` so that we only have to change code in one place if we alter how comps are disconnected in future

Shout if that all sounds ok!

Reply:
Yup, I believe your approach works. Now just replace `comps_to_disconnect.each` by `comps_to_disconnect.find_each` (see above for explanation) and I shall approve :D

Reply:
Oh and also `.where(end_date: ...DELAY_IN_DAYS.days.ago)` :wink:

Reply:
- `find_each` - made the change, plus a query on Slack
- What change are you wanting in the `where` clause? the `end_date:` looks like you want me to specify a parameter hash, but that doesn't seem possible with a `<`?

Reply:
My point is precisely that it _is_ possible!

- `Competition.where(end_date: 21.days.ago)` loads all competitions where end_date is equal to the specified value
- `Competition.where(end_date: ...21.days.ago)` basically does the same thing (load comps where the `end_date` column is equal to a certain value) but actually in this case the value you're passing is a Range object `a...b`. We don't know the start of the range, so we specify an open-ended exclusive range `...b` which means "anything up to b"

Reply:
Ooooh I thought that ellipsis was a rhetorical flourish. Makes a lot more sense! :D

Reply:
Yep works nicely! Latest commit includes the change.